<a href="https://colab.research.google.com/github/enmwmak/Teaching/blob/main/EIE4122/RNN_SentimentAnalysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **This Colab is based on the Colab file in [link](https://github.com/bentrevett/pytorch-sentiment-analysis/blob/master/1%20-%20Simple%20Sentiment%20Analysis.ipynb). To reduce running time, we will use a reduced-size IMDb dataset.**

# **Simple Sentiment Analysis**

We will be building a machine learning model to detect sentiment (i.e., detecting if a sentence is positive or negative) using PyTorch and TorchText. We will use a reduced-size [IMDb dataset](http://ai.stanford.edu/~amaas/data/sentiment/).

We will start with a simple model (RNN) to understand the general concepts instead of achieving good performance. In the next Colab, we will build on this knowledge to use more advanced models (Bi-directional LSTM) to obtain better results.

### **1. Introduction**

We'll be using a **recurrent neural network** (RNN) as they are commonly used in analysing sequences. An RNN takes in sequence of words, $X=\{x_1, ..., x_T\}$, one at a time, and produces a _hidden state_, $h$, for each word. We use the RNN _recurrently_ by feeding in the current word $x_t$ as well as the hidden state from the previous word, $h_{t-1}$, to produce the next hidden state, $h_t$. 

$$h_t = \text{RNN}(x_t, h_{t-1})$$

Once we have our final hidden state $h_T$ (from feeding in the last word in the sequence $x_T$), we feed $h_T$ through a fully-connected linear layer, followed by a sigmoid function $\sigma()$ to predict the sentiment score $\hat{y} = \sigma({\bf w}^\top_oh_T)$, where ${\bf w}_o$ contains the weights of the fully-connected layer at the output. Note that in the program "imdb-rnn.py", the sigmoid function is implemented in the function `binary_accuracy()`.

The following diagram shows an example sentence, with the RNN predicting zero, which indicates a negative sentiment. The unfolded RNN is shown in orange and the linear layer is shown in silver. Note that we use the same RNN for every word, i.e., it has the same parameters. The initial hidden state, $h_0$, is a tensor initialized to all zeros. 

![](https://github.com/bentrevett/pytorch-sentiment-analysis/blob/master/assets/sentiment1.png?raw=1)

**Note:** some layers and steps have been omitted from the diagram, but these will be explained later.

### **2. Preparing Data**

One of the main concepts of TorchText is the `Field`. These define how your data should be processed. In our sentiment classification task, the data consists of both the raw string of the review and the sentiment, either "pos" or "neg".

The parameters of a `Field` specify how the data should be processed. We use the `TEXT` field to define how the movie reviews should be processed, and the `LABEL` field determines how the labels should be processed. 

In "imdb-rnn.py", the `TEXT` field uses `tokenize='spacy'` as an argument. This specifies that "tokenization" (the act of splitting the string into discrete "tokens") should be done using the [spaCy](https://spacy.io) tokenizer. If no `tokenize` argument is passed, the default is simply splitting the string on spaces. We also need to specify a `tokenizer_language`, which tells TorchText which spaCy model to use. We use the `en_core_web_sm` model, which has to be downloaded with `python -m spacy download en_core_web_sm` before you run this notebook!

`LABEL` is defined by a `LabelField`, a special subset of the `Field` class specifically used for handling labels. We will explain the `dtype` argument later.

For more on `Fields`, go [here](https://pytorch.org/text/_modules/torchtext/data/field.html).

We also set the random seeds for reproducibility.

In [1]:
# Mount Google Drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
# Create a working folder and cd to it.
!mkdir -p /content/drive/MyDrive/Learning/EIE4122/lab2
%cd /content/drive/MyDrive/Learning/EIE4122/lab2

/content/drive/MyDrive/Learning/EIE4122/lab2


In [3]:
# Require gdown to download files from Google Drive
!pip install gdown

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [5]:
# Download the reduced-size IMDB dataset
!if [ ! -f data.tgz ]; then gdown https://drive.google.com/uc?id=1EL6YlIEs1IbUDjrZvrZ_9-C0C4gYoryZ; fi

Downloading...
From: https://drive.google.com/uc?id=1EL6YlIEs1IbUDjrZvrZ_9-C0C4gYoryZ
To: /content/drive/MyDrive/Learning/EIE4122/lab2/data.tgz
100% 67.8M/67.8M [00:00<00:00, 84.6MB/s]


In [6]:
!echo "Start decompression. It will take few minutes"
!tar zxf data.tgz --exclude="._*"

Start decompression. It will take few minutes


In [7]:
# Make sure that decompression is successful. The command "du" lists the size of folders.
!du -h .data/imdb/aclImdb

2.0M	.data/imdb/aclImdb/test/neg
2.2M	.data/imdb/aclImdb/test/pos
25M	.data/imdb/aclImdb/test
2.1M	.data/imdb/aclImdb/train/neg
2.3M	.data/imdb/aclImdb/train/pos
68M	.data/imdb/aclImdb/train
94M	.data/imdb/aclImdb


In [8]:
# Download torchtext. It will take a few minutes
!pip install torchtext==0.11.0

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 8.0 MB 27.6 MB/s 
     |██████████████████████████████▎ | 834.1 MB 1.2 MB/s eta 0:00:40tcmalloc: large alloc 1147494400 bytes == 0x38e9e000 @  0x7f023a905615 0x58ead6 0x4f355e 0x4d222f 0x51041f 0x5b4ee6 0x58ff2e 0x510325 0x5b4ee6 0x58ff2e 0x50d482 0x4d00fb 0x50cb8d 0x4d00fb 0x50cb8d 0x4d00fb 0x50cb8d 0x4bac0a 0x538a76 0x590ae5 0x510280 0x5b4ee6 0x58ff2e 0x50d482 0x5b4ee6 0x58ff2e 0x50c4fc 0x58fd37 0x50ca37 0x5b4ee6 0x58ff2e
     |████████████████████████████████| 881.9 MB 19 kB/s 
  Attempting uninstall: torch
    Found existing installation: torch 1.12.1+cu113
    Uninstalling torch-1.12.1+cu113:
      Successfully uninstalled torch-1.12.1+cu113
  Attempting uninstall: torchtext
    Found existing installation: torchtext 0.13.1
    Uninstalling torchtext-0.13.1:
      Successfully uninstalled torchtext-0.13.1
ERROR: pip's dependency resolver does n

In [9]:
# Define input TEXT and output LABEL
import torch
from torchtext.legacy import data

SEED = 1234
torch.manual_seed(SEED)
torch.backends.cudnn.deterministic = True
TEXT = data.Field(tokenize = 'spacy',
                  tokenizer_language = 'en_core_web_sm')
LABEL = data.LabelField(dtype = torch.float)

Another handy feature of TorchText is that it has support for common datasets used in natural language processing (NLP). The following code checks whether the IMDb dataset and has been downloaded to '.data/imdb/aclImdb/'. If yes, the code splits the dataset into the canonical train/test splits as `torchtext.datasets` objects. Otherwise, it will automatically download the original dataset from the internet. It processes the data using the `Fields` we have previously defined. 

The original IMDb dataset consists of 50,000 movie reviews, each marked as being a positive or negative review. However, to speed up training, we will only use 5,560 movie reviews in the original dataset.



In [10]:
# Because data have been uploaded to ".data/imdb/aclImdb/". This step will only take 1 minute.
# This step should be quick as PyTorch only need to know the structure of the data in the
# folder ".data/imdb/aclImdb". It will not actually load the data, which will be done by
# the Dataloader object (see code below)
from torchtext.legacy import datasets
train_data, test_data = datasets.IMDB.splits(TEXT, LABEL, path='.data/imdb/aclImdb/')

We can see how many examples are in each split by checking their length.

In [11]:
print(f'Number of training examples: {len(train_data)}')
print(f'Number of testing examples: {len(test_data)}')

Number of training examples: 2780
Number of testing examples: 2780


We can also check an example.

In [12]:
print(vars(train_data.examples[0]))

{'text': ['Wow', '!', 'Wow', '!', 'Wow', '!', 'I', 'have', 'never', 'seen', 'a', 'non', '-', 'preachy', 'documentary', 'on', 'globalization', 'until', 'I', 'saw', 'MARDI', 'GRAS', ':', 'MADE', 'IN', 'CHINA', '.', 'This', 'film', 'has', 'zero', 'narration', 'and', 'combines', 'verite', 'footage', 'with', 'sensitive', 'interviews', 'with', 'four', 'teenage', 'workers', 'in', 'China', 'who', 'live', 'inside', 'a', 'factory', 'compound', '.', 'They', 'play', 'with', 'toys', ',', 'jump', 'rope', ',', 'and', 'dance', '.', 'Yet', ',', 'the', 'majority', 'of', 'their', 'days', 'and', 'nights', 'consist', 'of', 'work', ',', 'work', ',', 'and', 'work', '--', 'but', 'the', 'footage', 'of', 'their', 'work', 'is', 'illuminating', 'and', 'mesmerizing', 'to', 'watch', '.', 'The', 'owner', 'of', 'the', 'factory', 'in', 'China', 'is', 'amazingly', 'open', ',', 'so', 'much', 'so', 'that', 'he', 'hits', 'home', 'the', 'effects', 'of', 'globalization', 'while', 'he', '"', 'punishes', '"', 'the', 'workers'

The IMDb dataset only has train/test splits, so we need to create a validation set. We can do this with the `.split()` method. 

By default this splits data into 70/30. However, by passing a `split_ratio` argument, we can change the split ratio, i.e., a `split_ratio` of 0.8 would mean 80% of the examples make up the training set and 20% make up the validation set. 

We also pass our random seed to the `random_state` argument to ensure that we get the same train/validation split each time.

In [13]:
import random
train_data, valid_data = train_data.split(random_state = random.seed(SEED))

Again, we can see how many examples are in each split.

In [14]:
print(f'Number of training examples: {len(train_data)}')
print(f'Number of validation examples: {len(valid_data)}')
print(f'Number of testing examples: {len(test_data)}')

Number of training examples: 1946
Number of validation examples: 834
Number of testing examples: 2780


Next, we have to build a _vocabulary_. This is effectively a look up table where every unique word in our data set has a corresponding _index_ (an integer). We need to do this because our machine learning model cannot operate on strings, only numbers. 

Each _index_ is used to construct a _one-hot_ vector for each word. A one-hot vector is a vector where all of the elements are 0, except one, which is 1, and the dimensionality is the total number of unique words in our vocabulary, commonly denoted by $V$.

![](https://github.com/bentrevett/pytorch-sentiment-analysis/blob/master/assets/sentiment5.png?raw=1)

The number of unique words in our training set is over 100,000, which means that our one-hot vectors will have over 100,000 dimensions! This will make training slow and possibly won't fit into our GPU. There are two effective ways to cut down our vocabulary: (1) only take the top $n$ most common words and (2) ignore words that appear less than $m$ times. We'll use the former, i.e., only keeping the top 25,000 words.

What do we do with words that are ignored? We may replace them with a special _unknown_ `<unk>` token. For example, if the sentence was "This film is great and I love it" but the word "love" was not in the vocabulary, it would become "This film is great and I `<unk>` it".

The following builds the vocabulary, only keeping the most common `max_size` tokens.

In [17]:
MAX_VOCAB_SIZE = 25_000    # In Python 25_000 means 25000
TEXT.build_vocab(train_data, max_size = MAX_VOCAB_SIZE)
LABEL.build_vocab(train_data)

Why do we only build the vocabulary on the training set? When testing any machine learning system you do not want to look at the test set in any way. We do not include the validation set as we want it to reflect the test set as much as possible.

In [18]:
print(f"Unique tokens in TEXT vocabulary: {len(TEXT.vocab)}")
print(f"Unique tokens in LABEL vocabulary: {len(LABEL.vocab)}")

Unique tokens in TEXT vocabulary: 25002
Unique tokens in LABEL vocabulary: 2


Why is the vocab size 25002 and not 25000? One of the addition tokens is the `<unk>` token and the other is a `<pad>` token.

When we feed sentences into our model, we feed a _batch_ of them at a time, i.e., more than one at a time, and all sentences in the batch need to be of the same size. Thus, to ensure that each sentence in the batch is of the same size, any sentences shorter than the longest within the batch are padded.

![](https://github.com/bentrevett/pytorch-sentiment-analysis/blob/master/assets/sentiment6.png?raw=1)

We can also view the most common words in the vocabulary and their frequencies.

In [19]:
print(TEXT.vocab.freqs.most_common(20))

[('the', 21892), (',', 21224), ('.', 18358), ('a', 12304), ('and', 12252), ('of', 10946), ('to', 10386), ('is', 8618), ('in', 6804), ('I', 5771), ('it', 5732), ('that', 5346), ('"', 5135), ("'s", 4792), ('this', 4555), ('-', 4282), ('was', 3872), ('/><br', 3819), ('as', 3400), ('with', 3357)]


We can also see the vocabulary directly using either the `stoi` (**s**tring **to** **i**nt) or `itos` (**i**nt **to**  **s**tring) method.

In [20]:
print(TEXT.vocab.itos[:10])      # Print from the beginning (0th) to the 9th position

['<unk>', '<pad>', 'the', ',', '.', 'a', 'and', 'of', 'to', 'is']


We can also check the labels, ensuring 0 is for negative and 1 is for positive.

In [21]:
print(LABEL.vocab.stoi)
print(LABEL.vocab.stoi['neg'])

defaultdict(None, {'neg': 0, 'pos': 1})
0


The final step of preparing the data is creating the iterators. We use the iterators to iterate over the training/validation sets in the training/evaluation loop, and they return a batch of examples (indexed and converted into tensors) at each iteration.

We'll use a `BucketIterator` which is a special type of iterator that will return a batch of examples where each example is of a similar length. This can minimize the amount of padding per example.

We also want to place the tensors returned by the iterator on the GPU (if you're using one). PyTorch handles this using `torch.device`. We then pass this device to the iterator.

In [22]:
BATCH_SIZE = 64
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
train_iterator, valid_iterator, test_iterator = data.BucketIterator.splits(
    (train_data, valid_data, test_data), 
    batch_size = BATCH_SIZE,
    device = device)

### **3. Build the Model**

The next stage is building the model that we'll eventually train and evaluate. There is a small amount of boilerplate code when creating models in PyTorch. Note that our `RNN` class is a sub-class of `nn.Module` and that we use `super` to call the superclass's `__init__` method.

Within the `__init__` we define the _layers_ of the module. The three layers comprises an _embedding_ layer, an _RNN_ layer, and a _linear_ layer. All layers have their parameters initialized to random values, unless explicitly specified.

The embedding layer is used to transform our sparse one-hot vector (sparse as most of the elements are 0) into a dense embedding vector (dense as the dimensionality is a lot smaller and all the elements are real numbers). This embedding layer is simply a single fully connected layer. In addition to reducing the dimensionality of the input to the RNN layer, there is the theory that words which have similar impact on the sentiment of the review are mapped close together in this dense vector space. For more information about word embeddings, see [here](https://monkeylearn.com/blog/word-embeddings-transform-text-numbers/).

The RNN layer takes in our dense vector and the previous hidden state $h_{t-1}$, which it uses to calculate the next hidden state, $h_t$.

![](https://github.com/bentrevett/pytorch-sentiment-analysis/blob/master/assets/sentiment7.png?raw=1)

Finally, the linear layer takes the final hidden state and feeds it through a fully-connected layer ${\bf w}_0^\top h_T$, transforming it to the correct output dimension. In our case, the output dimension is 1. The `forward` method is called when we feed examples into our model.

Each batch, `text`, is a tensor of size _**[sentence length, batch size]**_. That is a batch of sentences, each having each word converted into a one-hot vector. 

You may notice that this tensor should have another dimension due to the one-hot vectors, however PyTorch conveniently stores a one-hot vector as it's index value, i.e., the tensor representing a sentence is just a tensor of the indexes for each token in that sentence. The act of converting a list of tokens into a list of indexes is commonly called *numericalizing*.

The input batch is then passed through the embedding layer to get `embedded`, which gives us a dense vector representation of our sentences. `embedded` is a tensor of size _**[sentence length, batch size, embedding dim]**_.

`embedded` is then fed into the RNN layer. In some frameworks you must feed the initial hidden state, $h_0$, into the RNN, however in PyTorch, if no initial hidden state is passed as an argument it defaults to a tensor of all zeros.

The RNN returns 2 tensors, `output` of size _**[sentence length, batch size, hidden dim]**_ and `hidden` of size _**[1, batch size, hidden dim]**_. `output` is the concatenation of the hidden state from every time step, whereas `hidden` is simply the final hidden state. We verify this using the `assert` statement. Note the `squeeze` method, which is used to remove a dimension of size 1. 

Finally, we feed the last hidden state, `hidden`, through the linear layer, `fc`, to produce a prediction.

In [23]:
import torch.nn as nn

class RNN(nn.Module):
    def __init__(self, input_dim, embedding_dim, hidden_dim, output_dim):
        
        super().__init__()
        self.embedding = nn.Embedding(input_dim, embedding_dim)        
        self.rnn = nn.RNN(embedding_dim, hidden_dim)      
        self.fc = nn.Linear(hidden_dim, output_dim)
        
    def forward(self, text):
        #text = [sent len, batch size]        
        
        embedded = self.embedding(text)        
        #embedded = [sent len, batch size, emb dim]
        
        output, hidden = self.rnn(embedded)        
        #output = [sent len, batch size, hid dim]
        #hidden = [1, batch size, hid dim]
        assert torch.equal(output[-1,:,:], hidden.squeeze(0))
        
        return self.fc(hidden.squeeze(0))  # Remove the first dim in hidden to return [batch_size, hid_dim]

We now create an instance of our RNN class. The input dimension is the dimension of the one-hot vectors, which is equal to the vocabulary size. The embedding dimension is the size of the dense word vectors. This is usually around 50-250 dimensions, but depends on the size of the vocabulary.
The hidden dimension is the size of the hidden states. This is usually around 100-500 dimensions, but also depends on factors such as the vocabulary size, the size of the dense vectors and the complexity of the task. The output dimension is usually the number of classes, however in the case of only 2 classes the output value is between 0 and 1 and thus can be 1-dimensional, i.e., a single scalar real number.

In [24]:
INPUT_DIM = len(TEXT.vocab)
EMBEDDING_DIM = 100
HIDDEN_DIM = 256
OUTPUT_DIM = 1

model = RNN(INPUT_DIM, EMBEDDING_DIM, HIDDEN_DIM, OUTPUT_DIM)

Let's also create a function that will tell us how many trainable parameters our model has so we can compare the number of parameters across different models.

In [25]:
# Count the number of trainable parameters in our model
def count_parameters(model):
    return sum(p.numel() for p in model.parameters() if p.requires_grad)

print(f'The model has {count_parameters(model):,} trainable parameters')

The model has 2,592,105 trainable parameters


### **4. Train the Model**
Now we'll set up the training and then train the model. First, we'll create an optimizer. This is the algorithm we use to update the parameters of the module. Here, we'll use _stochastic gradient descent_ (SGD). The first argument is the parameters that will be updated by the optimizer. The second is the learning rate, i.e., how much we'll change the parameters when we do a parameter update.

In [26]:
import torch.optim as optim
optimizer = optim.SGD(model.parameters(), lr=1e-3)

Next, we'll define our loss function. In PyTorch, this is commonly called a criterion. The loss function here is _binary cross entropy with logits_. 

Our model currently outputs an unbound real number. As our labels are either 0 or 1, we want to restrict the predictions to a number between 0 and 1. We do this using the _sigmoid_ or _logit_ functions. We then use this this bound scalar to calculate the loss using binary cross entropy. The `BCEWithLogitsLoss` criterion carries out both the sigmoid and the binary cross entropy steps.

In [28]:
criterion = nn.BCEWithLogitsLoss()

Using `.to`, we can place the model and the criterion on the GPU (if we have one). 
<font color='red'>If this step takes a long period. Re-execute the statements may help.</font>

In [29]:
model = model.to(device)
criterion = criterion.to(device)

Our criterion function calculates the loss. However, we have to write our function to calculate the accuracy. This function first feeds the predictions through a sigmoid layer to squash the values between 0 and 1. Then, we round them to the nearest integer, i.e., rounding any value greater than 0.5 to 1 (a positive sentiment) and the rest to 0 (a negative sentiment). We then calculate how many rounded predictions equal the actual labels and average it across the batch.

In [34]:
def binary_accuracy(preds, y):
    """
    Returns accuracy per batch, i.e. if you get 8/10 right, this returns 0.8, NOT 8
    """
    #round predictions to the closest integer
    rounded_preds = torch.round(torch.sigmoid(preds))
    correct = (rounded_preds == y).float() #convert into float for division 
    acc = correct.sum() / len(correct)
    return acc

The `train` function iterates over all examples, one batch at a time. `model.train()` is used to put the model in "training mode", which turns on _dropout_ and _batch normalization_. Although we aren't using them in this model, it's good practice to include it.

For each batch, we first zero the gradients. Each parameter in a model has a `grad` attribute which stores the gradient calculated by the `criterion`. PyTorch does not automatically remove (or "zero") the gradients calculated from the last gradient calculation, so they must be manually zeroed.

We then feed the batch of sentences, `batch.text`, into the model. Note, you do not need to do `model.forward(batch.text)`. Instead, simply calling the model will do. The `squeeze` is needed as the predictions are initially size _**[batch size, 1]**_, and we need to remove the dimension of size 1 as PyTorch expects the predictions input to our criterion function to be of size _**[batch size]**_.

The loss and accuracy are then calculated using our predictions and the labels, `batch.label`, with the loss being averaged over all examples in the batch.

We calculate the gradient of each parameter with `loss.backward()`, and then update the parameters using the gradients and optimizer algorithm with `optimizer.step()`.
The loss and accuracy is accumulated across the epoch, the `.item()` method is used to extract a scalar from a tensor which only contains a single value. Finally, we return the loss and accuracy, averaged across the epoch. The `len` of an iterator is the number of batches in the iterator.

You may recall when initializing the `LABEL` field, we set `dtype=torch.float`. This is because TorchText sets tensors to be `LongTensor` by default, however our criterion expects both inputs to be `FloatTensor`. Setting the `dtype` to be `torch.float`, did this for us. The alternative method of doing this would be to do the conversion inside the `train` function by passing `batch.label.float()` instad of `batch.label` to the criterion. 

In [30]:
def train(model, iterator, optimizer, criterion, max_samples=2500):
    
    epoch_loss = 0
    epoch_acc = 0
    
    model.train()
    #for batch in iterator:
    for i, batch in enumerate(iterator):
        if i > max_samples:                               # Use less samples to reduce training time
            break  
        optimizer.zero_grad()                             # Initialize gradient to 0 for current batch
        predictions = model(batch.text).squeeze(1)        # Call RNN.forward() and get return object
        loss = criterion(predictions, batch.label)        # Compute binary cross-entropy
        acc = binary_accuracy(predictions, batch.label)
        loss.backward()                                   # Compute gradient
        optimizer.step()                                  # Update the weights by gradient descent
        epoch_loss += loss.item()
        epoch_acc += acc.item()
        
    return epoch_loss / len(iterator), epoch_acc / len(iterator)

`evaluate` is similar to `train`, with a few modifications as you don't want to update the parameters when evaluating.
`model.eval()` puts the model in "evaluation mode", which turns off _dropout_ and _batch normalization_. Again, we are not using them in this model, but it is good practice to include them.

No gradients are calculated on PyTorch operations inside the `with no_grad()` block. This causes less memory to be used and speeds up computation.

The rest of the function is the same as `train`, with the removal of `optimizer.zero_grad()`, `loss.backward()` and `optimizer.step()`, as we do not update the model's parameters when evaluating.

In [31]:
def evaluate(model, iterator, criterion, max_samples=2500):
    
    epoch_loss = 0
    epoch_acc = 0
    
    model.eval()
    with torch.no_grad():                                    # No gradient will be needed
        for batch in iterator:
            predictions = model(batch.text).squeeze(1)       # Call RNN.forward()           
            loss = criterion(predictions, batch.label)       # Compute binary cross entropy loss
            acc = binary_accuracy(predictions, batch.label)
            epoch_loss += loss.item()
            epoch_acc += acc.item()
        
    return epoch_loss / len(iterator), epoch_acc / len(iterator)

We'll also create a function to tell us how long an epoch takes to compare training times between models.

In [32]:
import time

def epoch_time(start_time, end_time):
    elapsed_time = end_time - start_time
    elapsed_mins = int(elapsed_time / 60)
    elapsed_secs = int(elapsed_time - (elapsed_mins * 60))
    return elapsed_mins, elapsed_secs

We then train the model through multiple epochs, an epoch being a complete pass through all examples in the training and validation sets. At each epoch, if the validation loss is the best we have seen so far, we'll save the parameters of the model and then after training has finished we'll use that model on the test set.

In [35]:
N_EPOCHS = 5
best_valid_loss = float('inf')
for epoch in range(N_EPOCHS):
    start_time = time.time()
    train_loss, train_acc = train(model, train_iterator, optimizer, criterion)
    valid_loss, valid_acc = evaluate(model, valid_iterator, criterion)
    end_time = time.time()
    epoch_mins, epoch_secs = epoch_time(start_time, end_time)
    
    if valid_loss < best_valid_loss:
        best_valid_loss = valid_loss
        torch.save(model.state_dict(), 'tut1-model.pt')
    
    print(f'Epoch: {epoch+1:02} | Epoch Time: {epoch_mins}m {epoch_secs}s')
    print(f'\tTrain Loss: {train_loss:.3f} | Train Acc: {train_acc*100:.2f}%')
    print(f'\t Val. Loss: {valid_loss:.3f} |  Val. Acc: {valid_acc*100:.2f}%')

Epoch: 01 | Epoch Time: 0m 2s
	Train Loss: 0.697 | Train Acc: 49.65%
	 Val. Loss: 0.695 |  Val. Acc: 50.33%
Epoch: 02 | Epoch Time: 0m 1s
	Train Loss: 0.694 | Train Acc: 49.52%
	 Val. Loss: 0.695 |  Val. Acc: 50.33%
Epoch: 03 | Epoch Time: 0m 1s
	Train Loss: 0.694 | Train Acc: 48.81%
	 Val. Loss: 0.696 |  Val. Acc: 50.33%
Epoch: 04 | Epoch Time: 0m 1s
	Train Loss: 0.693 | Train Acc: 48.91%
	 Val. Loss: 0.696 |  Val. Acc: 50.45%
Epoch: 05 | Epoch Time: 0m 1s
	Train Loss: 0.693 | Train Acc: 49.58%
	 Val. Loss: 0.696 |  Val. Acc: 45.09%


You may have noticed the loss is not really decreasing and the accuracy is poor. This is due to several issues with the model which we'll improve in the next notebook.

Finally, the metric we actually care about, the test loss and accuracy, which we get from our parameters that gave us the best validation loss.

In [36]:
model.load_state_dict(torch.load('tut1-model.pt'))
test_loss, test_acc = evaluate(model, test_iterator, criterion)
print(f'Test Loss: {test_loss:.3f} | Test Acc: {test_acc*100:.2f}%')

Test Loss: 0.698 | Test Acc: 49.36%


### **5. Further Investigations**

Try the following:
1.   Increase the number of RNN layers 
2.   Change the embedding dimension of your RNN
3.   Reduce the dimension of word embedding

Discuss your results and observations in your report. Also, determine the type of loss function used by the RNN in this program.